integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,92 @@
+ */
+val Lifecycle.coroutineScope: CoroutineScope
+    get() = lifecycleCoroutineScopes[this] ?: job.let { job ->
+        val newScope = CoroutineScope(job + Dispatchers.Main)
+        if (job.isActive) {
+            lifecycleCoroutineScopes[this] = newScope
+            job.invokeOnCompletion { _ -> lifecycleCoroutineScopes -= this }
+        }
+        newScope
+    }
+
+/**
+ * Calls [Lifecycle.coroutineScope] for the [Lifecycle] of this [LifecycleOwner].
+ *
+ * This is an inline property, just there for convenient usage from any [LifecycleOwner],
+ * like FragmentActivity, AppCompatActivity, Fragment and LifecycleService.
+ */
+inline val LifecycleOwner.coroutineScope get() = lifecycle.coroutineScope
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that will be cancelled as
+ * soon as this [Lifecycle] [currentState][Lifecycle.getCurrentState] is no longer
+ * [at least][Lifecycle.State.isAtLeast] the passed [activeWhile] state.
+ *
+ * **Beware**: if the current state is lower than the passed [activeWhile] state, you'll get an
+ * already cancelled scope.
+ */
+fun Lifecycle.createScope(activeWhile: Lifecycle.State): CoroutineScope {
+    if (activeWhile == DESTROYED) return coroutineScope
+    return CoroutineScope(createJob(activeWhile) + Dispatchers.Main)
+}
+
+/**
+ * Returns a [Job] that will be cancelled as soon as the [Lifecycle] reaches
+ * [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ *
+ * You can use this job for custom [CoroutineScope]s, or as a parent [Job].
+ */
+val Lifecycle.job: Job
+    get() = lifecycleJobs[this] ?: createJob().also {
+        if (it.isActive) {
+            lifecycleJobs[this] = it
+            it.invokeOnCompletion { _ -> lifecycleJobs -= this }
+        }
+    }
+
+/**
+ * Creates a [Job] that will be cancelled as soon as this [Lifecycle]
+ * [currentState][Lifecycle.getCurrentState] is no longer [at least][Lifecycle.State.isAtLeast] the
+ * passed [activeWhile] state.
+ *
+ * **Beware**: if the current state is lower than the passed [activeWhile] state, you'll get an
+ * already cancelled job.
+ */
+fun Lifecycle.createJob(activeWhile: Lifecycle.State = INITIALIZED): Job {
+    require(activeWhile != Lifecycle.State.DESTROYED) {
+        "DESTROYED is a terminal state that is forbidden for createJob(…), to avoid leaks."
+    }
+    return Job().also { job ->
+        if (!currentState.isAtLeast(activeWhile)) job.cancel()
+        else addObserver(object : GenericLifecycleObserver {
+            override fun onStateChanged(source: LifecycleOwner?, event: Lifecycle.Event) {
+                if (!currentState.isAtLeast(activeWhile)) {
+                    removeObserver(this)
+                    job.cancel()
+                }
+            }
+        })
+    }
+}
+
+private val lifecycleJobs = ConcurrentHashMap<Lifecycle, Job>()
```

Comment:
I don't think an entry in the `ConcurrentHashMap` is so expensive, especially when compared to the cost of creating everything associated to the `LifecycleOwner` (like UI in an `Activity`).

If you want to cache the scope the most efficient way, you can still implement `CoroutineScope` and implement it this way: `override coroutineContext = lifecycle.coroutineContext` to cache without map access, but by default, you don't need to do it at all.

The reason I did it is to have property syntax (where repeated access doesn't mean repeated allocation), and promote lifecycle scoped usage.

I'm wondering if `mutableMap`, instead of `ConcurrenHashMap` would still work correctly given the `MutableMap` is only used for caching, and that removal from the cache doesn't rely on the map itself, but just on the observer. I think it would be safe for this usage. Anyone thinks otherwise?

Also @gildor, with this information about the why, do you still think `coroutineScope` extension properties should be removed? I personally use it very often. I can make a separate library just for these two extensions on `Lifecycle` and `LifecycleOwner`, but that doesn't seem to be a bad API to me, so I think it could be good to have it into this integration by default. Let me know your rationale or what you think about it :)

Reply:
> I don't think an entry in the ConcurrentHashMap is so expensive, especially when compared to the cost of creating everything associated to the LifecycleOwner (like UI in an Activity).

Why do you compare the creation of Activity vs ConcurrentHashMap instead of a comparison of ConcurrentHashMap vs just a field to cache scope?
I think an extension for creating scope from lifecycle is completely fine, but it should create scope only, like createJob() instead of an implicit caching instance of scope to a global static map.
My point that this is performance optimization with unclear performance gain, imagine if you have some component that uses scope created from lifecycle for example 3-5 times.
You have 3 choices:
1. Create a scope, cache to a field of your component
2. Do not cache scope, so will be created 3 scope instances
3. Use this extension, create a scope, cache it to ConcurrentHashMap

the first option requires more code but obviously more performant (less object created, no atomic or synchronized operations).

Then you have a choice between 2 and 3. And this is not obvious for me that such caching on practice (or even on benchmark) really would be more performant.
So this is why I against this solution and believe that kotlinx.coroutines shouldn't encourage usage of this pattern because it's too implicit and doesn't provide clear advantages or performance gain.

Reply:
> Why do you compare the creation of Activity vs ConcurrentHashMap instead of a comparison of ConcurrentHashMap vs just a field to cache scope?

I'm comparing the runtime cost (memory and CPU), with same code amount, to reason about the slight overhead implied by the global caching map.

You can already decide to only use `createScope(…)` or `createJob(…)` and cache it in a property that you can name `coroutineScope` if you want (as members take precedence over extensions), or use it to implement `CoroutineScope`.

However, the reason why I provide the `coroutineScope` extension property is to allow you to use a good default (scope on `Dispatchers.Main` that is cancelled when `lifecycle` reaches `DESTROYED` state) without having to write any boilerplate.

If you start only one coroutine on this default scope for your `Lifecycle`, then you might as well use `createScope(…)` and keep the global caching map untouched.

Right now, these `coroutineScope` extension properties and their cache is in the same file and underlying class as the other extensions, but per your recommendation, I'll make it two classes, so if you don't use `coroutineScope`, then the global caching map is never initialized as its class is never loaded. I think even without this, if you don't use it, proguard removes it when run (e.g. on your release builds).

For the cache, I think `mutableMap` should be enough because it's just a cache, and it's not a big deal if the returned job or scope is not the one created on first access, as they will all behave the same way and receive the state updates anyway, in the case you get to this point with multithreaded access. I'll just give myself a bit more thinking time to be certain about this.

Reply:
@gildor I just pushed 4 commits to make use of `coroutineScope` optional (no longer referenced from `createScope`, put in a separate file), and replaced `ConcurrentHashMap` by `mutableMapOf` as we don't need the concurrency safety features for this caching use case.

Also, I made it clear, by the name, that the maps are for caching.

Please, let me know what you think of the current version.

Reply:
>and replaced ConcurrentHashMap by mutableMapOf as we don't need the concurrency safety features for this caching use case.

This is an illegal transformation. Now any access to this map may hang indefinitely (yes, it is possible) or throw NPE.

Reply:
I had a hard time believing this, and then read that because of the loops and hashing operations in `HashMap`, you can get an infinite loop on an attempt to read. Thanks for the information, TIL.

I'll revert my change back to `ConcurrentHashMap` to make it safe again, even though this may not make it into this repo, neither into AndroidX (as being inside AndroidX will likely allow an alternative implementation that doesn't need to rely on any global cache). This is to leave a clean PR in case anyone wants to integrate this into a project without waiting for any AndroidX built-in support.